### PR TITLE
Prepare to move test code to `Validation/ExecutionContext` API

### PR DESCRIPTION
**Pull request by plafer**
Work towards: #413
Work towards: #426

## Description

- Demonstrates how test code is refactored for the `timeout_on_close` handler only
  - New tests are written using the `rstest` crate
- Sets up next PR by moving all "ICS-26 router" functions to their own submodule to facilitate unit testing

Ultimate goal (finished in next PR) is to remove old `Reader/Keeper` context implementations from `MockContext`, which requires updating all tests to test `validate()` and `execute()` functions as opposed to `process()`.

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests.
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/src/mock/context.rs [hunk 1]
```diff
@@ -668,6 +679,7 @@ impl RouterBuilder for MockRouterBuilder {
     }
 }
 
+// TODO: REMOVE BEFORE MERGING
```
Comment (Farhad-Shabani) [inline crates/ibc/src/mock/context.rs]: any action needed for this comment?

File: crates/ibc/src/core/context.rs [hunk 2]
```diff
@@ -596,865 +601,3 @@ pub trait ExecutionContext: ValidationContext {
     /// Logging facility
     fn log_message(&mut self, message: String);
 }
-
-fn chan_open_init_validate<ValCtx>(
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/context.rs:600]: It got cleaned, perfect!

File: crates/ibc/src/core/handler.rs [hunk 3]
```diff
@@ -28,3 +28,684 @@ pub fn dispatch(ctx: &mut impl ExecutionContext, msg: MsgEnvelope) -> Result<(),
...
+    use crate::test_utils::{get_dummy_account_id, DummyTransferModule};
+    use crate::timestamp::Timestamp;
+    use crate::Height;
+
+    #[test]
+    /// These tests exercise two main paths: (1) the ability of the ICS26 routing module to dispatch
+    /// messages to the correct module handler, and more importantly: (2) the ability of ICS handlers
+    /// to work with the context and correctly store results (i.e., the `ClientKeeper`,
+    /// `ConnectionKeeper`, and `ChannelKeeper` traits).
+    fn routing_module_and_keepers() {
```
Comment (Farhad-Shabani) [inline crates/ibc/src/core/handler.rs:105]: Are we gonna also use `rstest` fixtures here later?

**Review by Farhad-Shabani — APPROVED** [review]
:ok_hand:

File: crates/ibc/src/core/handler.rs (hunk 3 above)
Comment (plafer) [reply crates/ibc/src/core/handler.rs:105]: I think there's a better way to write it using `rstest` yes; not sure exactly which feature to use. I'll open an issue for it.

File: crates/ibc/src/mock/context.rs (hunk 1 above)
Comment (plafer) [reply crates/ibc/src/mock/context.rs]: oops, that will be removed in next PR now. Removed comment: [5c602f1](https://github.com/cosmos/ibc-rs/pull/417/commits/5c602f1fd504c7bd7a7453f913a09359f2457800)

**Pull request merged**